### PR TITLE
Preserve chmod and chown from overwritten file

### PR DESCRIPTION
**Pull request by sheerun**
Among others it fixes:

- https://github.com/npm/write-file-atomic/issues/11
- https://github.com/yeoman/update-notifier/issues/76
- https://github.com/tessel/t2-cli/issues/727
- [bower permission issues](https://github.com/bower/bower/issues?q=is%3Aissue+configstore+is%3Aclosed)
- [f**kton](https://github.com/search?p=2&q=configstore+permission&ref=searchresults&type=Issues&utf8=%E2%9C%93) of similar ones

### CODE REVIEW

**Comment by sheerun** [comment]
@jimbojw @sindresorhus @Qix- could you test?

**Comment by sindresorhus** [comment]
// @LinusU @nexdrew @jephir

**Comment by sindresorhus** [comment]
> README.md 100644 → 100755

You probably didn't mean to change this.

**Comment by sheerun** [comment]
Hmm I need to fix sync version as well

**Comment by sheerun** [comment]
Everything should be ok now. I've also fixed chmod behavior on node 0.8 which was broken before (node 0.8 doesn't support mode for writeFile).

**Comment by nexdrew** [comment]
I have tested and confirmed that this fixes https://github.com/yeoman/update-notifier/issues/76. Nice work @sheerun!

**Comment by Qix-** [comment]
Looks like it worked 💃

**Comment by iarna** [comment]
Why the `sudo` testing? We'd very much like to NOT turn sudo on, as it means using Travis' legacy infrastructure.

File: index.js [hunk 1]
```diff
@@ -20,22 +20,68 @@ module.exports = function writeFile (filename, data, options, callback) {
...
+  function computeOptions (filename, cb) {
+    if (options.mode && options.chmod) {
+      // Fast-track, no need for fetching stat from previous file
+      cb(null)
+    } else {
+      // Either mode or chown is not explicitly set
+      // Default behavior is to copy it from original file
+      fs.stat(filename, function (err, stats) {
+        if (!err && stats && !options.mode) {
+          options.mode = stats.mode
```
Comment (iarna) [inline index.js]: I'm not comfortable mutating the options object folks pass in.

File: .travis.yml [hunk 2]
```diff
@@ -1,11 +1,21 @@
...
+script:
+  - "sudo $(which node) $(which npm) test"
+os:
+  - osx
+  - linux
+matrix:
+  fast_finish: true
+  allow_failures:
+    - os: osx
 node_js:
```
Comment (iarna) [inline .travis.yml]: I'd be happy to see the version list here updated to match npm itself:
> 4, 6, 0.10, 5, 0.12

in that order

**Comment by sheerun** [comment]
There's no other way of integration testing "chmod" without enabling sudo. Please try to disable it and observe the tests fail.

**Comment by sheerun** [comment]
All fixed

**Comment by jimbojw** [comment]
Thanks @sheerun!

**Comment by sheerun** [comment]
I've also fixed option overriding issue in sync version

**Comment by sheerun** [comment]
ping

**Comment by sindresorhus** [comment]
ping

**Comment by Qix-** [comment]
gnip

**Comment by jimbojw** [comment]
Linked some [related bugs](https://github.com/search?utf8=%E2%9C%93&q=%22configstore%22+%22permission%22+%22sudo+npm%22+%22eacces%22&type=Issues&ref=searchresults) to raise visibility.

**Comment by sheerun** [comment]
@iarna Any plans to merge it?

**Comment by iarna** [comment]
Yeah, that makes sense, I'll merge this shortly.

**Comment by iarna** [comment]
Merged as 36f5104 and released in v1.2.0

**Pull request closed without merging**